src/modules/decentralized-web-node/services/message.service.ts

```diff
@@ -0,0 +1,115 @@
+import { Injectable, UnauthorizedException } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { FindManyOptions, Repository } from 'typeorm';
+import { Logger } from '../../logger/logger.service';
+import { DataSort } from '../dtos/collections-query-descriptor.dto';
+import { Message } from '../entities/message.entity';
+
+@Injectable()
+export class MessageService {
+  constructor(
+    private readonly logger: Logger,
+    @InjectRepository(Message)
+    private readonly messageRepository: Repository<Message>
+  ) {
+    this.logger.setContext(MessageService.name);
+  }
+
+  async find(
+    user: string,
+    options: {
+      dateSort?: DataSort;
+      dataFormat?: string;
+      schema?: string;
+      objectId?: string;
+    }
+  ): Promise<Message[]> {
+    const whereParams: Record<string, string | number> = {
+      createdBy: user,
+    };
+
+    if (options.schema) {
+      whereParams.schema = options.schema;
+    }
+
+    if (options.dataFormat) {
+      whereParams.dataFormat = options.dataFormat;
+    }
+
+    if (options.objectId) {
+      whereParams.objectId = options.objectId;
+    }
+
+    let order: FindManyOptions<Message>['order'];
+    switch (options.dateSort) {
+      case 'createdDescending':
+        order = {
+          dateCreated: 'DESC',
+        };
+        break;
+      case 'publishedAscending':
+        order = {
+          datePublished: 'ASC',
+        };
+        break;
+      case 'publishedDescending':
+        order = {
+          datePublished: 'DESC',
+        };
+        break;
+      default:
+      case 'createdAscending':
+        order = {
+          dateCreated: 'ASC',
+        };
+        break;
+    }
+
+    const items = await this.messageRepository.find({
+      where: whereParams,
+      order,
+    });
+    return items;
+  }
+
+  async save(message: Message, user: string): Promise<Message> {
+    const foundedMessage = await this.messageRepository.findOne({
+      where: {
+        objectId: message.objectId,
+      },
+    });
+    let item: Message = foundedMessage;
+
+    if (foundedMessage && foundedMessage.createdBy !== user) {
+      throw new UnauthorizedException(
+        `User is not allowed to update message ${message.objectId}`
+      );
+    }
```

Comment:
Does still code mean that the implementation can't store a message with the same `objectId` but created by different users?
I think different users could store the same object, we should instead authorize at the `target` level

Reply:
ObjectId is the uuid v4 version, do you think it should be the same for all objects?

Reply:
Hmm, sorry, maybe I confused. But what I'm trying to say is that we should allow multiple `targets` to store the same object and I don't think this code allows this because it assumes that there is only one of the message with a given `objectId` (because it uses `findOne` and doesn't filter on `target`). Is this not how you see it?

Reply:
Anyone can create own `objectId` it can be different for the same message data

Reply:
Right, I'm imagining a kind of multicast messaging scenario where multiple targets (all using the same decentralized web node server) receive the same message (so it would have the same `objectId`). I think this code would fail in this case, right?

Reply:
Ohh, okay I see. let me fix that